crypto/pkcs12/p12_crt.c

```diff
@@ -69,12 +72,30 @@ PKCS12 *PKCS12_create_ex(const char *pass, const char *name, EVP_PKEY *pkey,
             goto err;
         if (keyidlen && !PKCS12_add_localkeyid(bag, keyid, keyidlen))
             goto err;
+        if (cb != NULL) {
+            cbret = cb(bag, cbarg);
+            if (cbret  == -1) {
+                ERR_raise(ERR_LIB_PKCS12, PKCS12_R_CALLBACK_FAILED);
+                goto err;
+            } else if (cbret == 0) {
+                PKCS12_remove_cert(&bags, bag);
+            }
+        }
     }
 
     /* Add all other certificates */
     for (i = 0; i < sk_X509_num(ca); i++) {
-        if (!PKCS12_add_cert(&bags, sk_X509_value(ca, i)))
+        if ((bag = PKCS12_add_cert(&bags, sk_X509_value(ca, i))) == NULL)
             goto err;
+        if (cb != NULL) {
+            cbret = cb(bag, cbarg);
+            if (cbret  == -1) {
+                ERR_raise(ERR_LIB_PKCS12, PKCS12_R_CALLBACK_FAILED);
+                goto err;
+            } else if (cbret == 0) {
+                PKCS12_remove_cert(&bags, bag);
+            }
+        }
     }
 
     if (bags && !PKCS12_add_safe_ex(&safes, bags, nid_cert, iter, pass,
```

Comment:
We should also call the callback for the bag containing the key as well. *Our* callback doesn't need that but other applications might.